File: backend/src/Host/Configuration/OwnerSettings/Administration/OwnDisplayName.cs
// Copyright © 2026 Krzysztof Kasprowicz
// Licensed under the GNU Affero General Public License, Version 3. See LICENSE in the project root for license information.
// Project repository: https://github.com/Krzysztof318/MailFathom

using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using MailFathom.Application.Access;
using MailFathom.Domain.Access;

namespace MailFathom.Host.Configuration.OwnerSettings.Administration;

/// <summary>Hands the signed-in person the name this deployment records them under, and takes back the one they correct it to.</summary>
/// <remarks>
/// <para>
/// The name is the envelope rather than the document — the column an operator tells one owner from another by, unique
/// across the deployment and keyed by nothing — which is why this stands beside
/// <see cref="OwnerRecordAdministration" /> instead of inside it. What it shares with that service is the gate: the
/// envelope is written under the record's own grant and refused for the same owner, because both are what this
/// deployment holds about a person rather than what that person set about their client.
/// </para>
/// <para>
/// <b>Neither act names an owner.</b> The person is the one the credential authenticated, resolved from the principal
/// exactly as the record's own acts resolve it, so a request about somebody else is something a caller cannot express.
/// The read is a key lookup on that one owner rather than a roster filtered down to them, which is what keeps an
/// owner-facing surface from composing a deployment-wide catalog of people.
/// </para>
/// <para>
/// Reading is <see cref="MailFathomPermission.MailRead" />, the grant a signed-in person already holds, because a
/// person who may not change their name must still be shown it. Writing is
/// <see cref="MailFathomPermission.MailAccountsWrite" />, which is the record's write and is granted separately — and
/// the read says whether a write would be accepted, so a client draws the name as text rather than discovering the
/// refusal by submitting one.
/// </para>
/// <para>
/// A person whose mail accounts a configuration source still declares is refused, for the reason the record's own
/// write refuses them: a start relabels every declared owner from the declaration, so a name written here would stand
/// until the next restart and then silently revert. What comes back names the file to correct instead.
/// </para>
/// </remarks>
[SuppressMessage(
    "Performance",
    "CA1812:Avoid uninstantiated internal classes",
    Justification = "The dependency injection container materializes this service.")]
internal sealed class OwnDisplayName(
    AccessAuthorization authorization,
    IMailOwnerDirectory directory,
    IMailOwnerProvisioning provisioning,
    ServedMailOwners servedOwners)
{
    /// <summary>Reads the name this deployment records the signed-in person under.</summary>
    /// <param name="cancellationToken">Cancels the read.</param>
    /// <returns>The name and whether this caller could change it, or <see langword="null" /> when this deployment holds no record for them.</returns>
    /// <exception cref="PrincipalNotAuthorizedException">Thrown when the caller acts for no owner, or its grant omits <see cref="MailFathomPermission.MailRead" />.</exception>
    internal async Task<OwnDisplayNameReading?> ReadAsync(CancellationToken cancellationToken)
    {
        authorization.RequirePermission(MailFathomPermission.MailRead);

        var owner = authorization.RequireOwner();

        return await directory.ReadOwnerAsync(owner, cancellationToken) is { } held
            ? new OwnDisplayNameReading(held.DisplayName, this.WouldAcceptAWriteFor(owner))
            : null;
    }

    /// <summary>Records the signed-in person under the name they corrected theirs to.</summary>
    /// <param name="displayName">The name they would be recorded under.</param>
    /// <param name="cancellationToken">Cancels the read and the write.</param>
    /// <returns>What the change did: whether this deployment holds them at all, the name the row now carries, and the sentence naming what to correct where it holds them and refused.</returns>
    /// <exception cref="PrincipalNotAuthorizedException">Thrown when the caller acts for no owner, or its grant omits <see cref="MailFathomPermission.MailAccountsWrite" />.</exception>
    /// <remarks>
    /// The version the record's writes state has no counterpart here, for the reason <c>RelabelAsync</c> gives: the
    /// version guards the document a change is composed over and the name is not part of that document. The
    /// uniqueness is guarded by the statement itself rather than by a roster read before it, so a name taken between
    /// this read and that write is a refusal rather than the server's own unique-violation sentence.
    /// </remarks>
    internal async Task<OwnDisplayNameChange> ChangeAsync(string? displayName, CancellationToken cancellationToken)
    {
        authorization.RequirePermission(MailFathomPermission.MailAccountsWrite);

        var owner = authorization.RequireOwner();

        if (await directory.ReadOwnerAsync(owner, cancellationToken) is null)
        {
            return OwnDisplayNameChange.NoSuchOwner;
        }

        // Ahead of the bound, exactly as the record's own gate is checked ahead of the version it was composed over:
        // a person a file still supplies is refused whatever name they wrote, and telling them to shorten one first
        // would send them back to a field that was never going to be accepted.
        if (servedOwners.SourceFor(owner) != MailOwnerAccountSource.OwnerDocument)
        {
            return OwnDisplayNameChange.Refused(DeclaredElsewhere);
        }

        if (FindNameRefusal(displayName) is { } unusable)
        {
            return OwnDisplayNameChange.Refused(unusable);
        }

        var name = displayName!.Trim();

        return await provisioning.RelabelAsync(owner, name, cancellationToken)
            ? OwnDisplayNameChange.Recording(name)
            : OwnDisplayNameChange.Refused(NameTaken);
    }

    /// <summary>Reports whether this caller could change the name of the owner they act for.</summary>
    /// <remarks>
    /// Both halves of the write's own gate, asked without attempting one: the grant the credential carries, and the
    /// source this deployment reads the person's mail accounts from. A caller holding one and not the other is
    /// answered the same as one holding neither, because what a client does about either is the same — draw the name
    /// and offer nothing to change it with.
    /// </remarks>
    private bool WouldAcceptAWriteFor(MailOwnerId owner) =>
        authorization.Permits(MailFathomPermission.MailAccountsWrite)
        && servedOwners.SourceFor(owner) == MailOwnerAccountSource.OwnerDocument;

    /// <summary>Says why a stated name is not one this deployment would record, or nothing where it is.</summary>
    /// <remarks>
    /// The rules are the ones an administrator's own relabel applies, because both write the same column; the
    /// sentences are not, because this one is read by the person being named rather than by whoever administers them.
    /// </remarks>
    private static string? FindNameRefusal(string? displayName)
    {
        if (string.IsNullOrWhiteSpace(displayName))
        {
            return "Your name is what this deployment records you as, so a change of it states one. Write a name, unique across this deployment.";
        }

        var name = displayName.Trim();

        return name.Length > MailOwnerRecord.MaximumDisplayNameLength
            ? string.Create(
                CultureInfo.InvariantCulture,
                $"The name is {name.Length} characters, past the {MailOwnerRecord.MaximumDisplayNameLength} this deployment stores. Shorten it.")
            : null;
    }

    /// <summary>The sentence a name another person of this deployment already carries is refused with.</summary>
    /// <remarks>It names no one: that the name is taken is what the person has to act on, and who took it is somebody else's record.</remarks>
    private const string NameTaken =
        "Somebody else on this deployment is already recorded under that name, and a name is unique across it. Choose another.";

    /// <summary>The sentence a person whose mail accounts a configuration source declares is refused with.</summary>
    /// <remarks>
    /// It names both shapes the declaration takes, for the reason the erasure's own refusal does: which one an
    /// operator wrote is in their file rather than in anything this deployment could report back to the person.
    /// </remarks>
    private const string DeclaredElsewhere =
        "A configuration source declares your mail accounts, so this deployment reads your name from it too and writes that name back at every start — a change made here would stand until the next restart and then revert. Ask whoever administers this deployment to change the name in your entry of the top-level Accounts collection, or to move your accounts into your own record with 'mfctl owner adopt'.";
}

File: frontend/src/Client.Backend/src/ownDisplayName.ts
// Copyright © 2026 Krzysztof Kasprowicz
// Licensed under the GNU Affero General Public License, Version 3. See LICENSE in the project root for license information.
// Project repository: https://github.com/Krzysztof318/MailFathom

import { failed, failureReasonForStatus, read, type ClientFailure, type ClientResult } from './failure';
import { asRecord } from './json';
import { headersFor, routeFor, type ClientSession } from './session';
import { spanned } from './telemetry';
import { send, type ClientResponse, type MailFathomTransport } from './transport';

// What the signed-in person is called, which is the one thing a client drawing them cannot compose from the credential
// it holds. The read says whether this deployment would take a correction, so a screen draws the field it will
// actually be allowed to write rather than finding out by being refused.

/** The route the acting person's own name is read at and written back to, relative to the client prefix. */
export const ownDisplayNameRoute = '/display-name';

/** The name this deployment records the signed-in person under, and whether they may correct it here. */
export interface OwnDisplayName {
    readonly displayName: string;

    /**
     * Whether a correction from this caller would be accepted.
     *
     * The deployment answers it rather than leaving a client to discover it, and it does not say which of the two
     * things refuses one — a grant the credential lacks, or mail accounts a configuration source still declares. A
     * screen draws the same read-only field either way.
     */
    readonly changeable: boolean;
}

/**
 * The most of one name answer this package reads before refusing it.
 *
 * One name bounded at 128 characters and one boolean, with room for the widest UTF-8 encoding of each character and
 * the JSON around them. It is the same order the write route bounds its request body at, and for the same reason: what
 * the bound guards against is an answer that was never a name.
 */
export const longestDisplayNameAnswer = 1_024;

/** Reads the name this deployment records the signed-in person under, answering an expected failure as a value. */
export function readOwnDisplayName(
    session: ClientSession,
    transport: MailFathomTransport,
): Promise<ClientResult<OwnDisplayName>> {
    return spanned(`GET ${ownDisplayNameRoute}`, async () => {
        const response = await send(transport, {
            method: 'GET',
            path: routeFor(session, ownDisplayNameRoute),
            headers: headersFor(session),
            longestAnswer: longestDisplayNameAnswer,
        });

        if (response === null) {
            return failed('unavailable', null);
        }

        if (response.status !== 200) {
            return failed(failureReasonForStatus(response.status), response.status);
        }

        const name = parseDisplayName(response.body);

        return name === null ? failed('unreadable', response.status) : read(name);
    });
}

/**
 * What became of a correction somebody made to their own name.
 *
 * A refused name is separated from every other failure because it is the one a person acts on: what they typed is
 * blank, longer than the deployment stores, or a name somebody else there already carries, and the answer is to type
 * another one rather than to sign in again or to try later. The four reasons say nothing that reaches that, which is
 * why this is an outcome of its own rather than a fifth member added to them.
 */
export type OwnDisplayNameChange =
    | { readonly outcome: 'recorded'; readonly displayName: string }
    | { readonly outcome: 'notAcceptable' }
    | { readonly outcome: 'failed'; readonly failure: ClientFailure };

/**
 * Records the signed-in person under the name they corrected theirs to.
 *
 * The answer carries the name as it was stored rather than as it was sent, because a name is trimmed on its way in and
 * a screen that redrew what was typed would show something the deployment does not hold.
 */
export async function changeOwnDisplayName(
    session: ClientSession,
    transport: MailFathomTransport,
    displayName: string,
): Promise<OwnDisplayNameChange> {
    const response = await send(transport, {
        method: 'POST',
        path: routeFor(session, ownDisplayNameRoute),
        headers: { ...headersFor(session), 'Content-Type': 'application/json' },
        body: JSON.stringify({ displayName }),
        longestAnswer: longestDisplayNameAnswer,
    });

    return changeOf(response);
}

function changeOf(response: ClientResponse | null): OwnDisplayNameChange {
    if (response === null) {
        return { outcome: 'failed', failure: { reason: 'unavailable', status: null } };
    }

    // The deployment names what to correct in the body it refuses with, and none of that reaches the screen: what it
    // says is about this deployment's own entries, and the sentence a person is shown is the client's to word.
    if (response.status === 400) {
        return { outcome: 'notAcceptable' };
    }

    if (response.status !== 200) {
        return {
            outcome: 'failed',
            failure: { reason: failureReasonForStatus(response.status), status: response.status },
        };
    }

    const recorded = parseDisplayName(response.body);

    return recorded === null
        ? { outcome: 'failed', failure: { reason: 'unreadable', status: response.status } }
        : { outcome: 'recorded', displayName: recorded.displayName };
}

function parseDisplayName(body: string): OwnDisplayName | null {
    let parsed: unknown;

    try {
        parsed = JSON.parse(body);
    } catch {
        return null;
    }

    const record = asRecord(parsed);
    if (record === null) {
        return null;
    }

    const displayName = record['displayName'];
    const changeable = record['changeable'];

    if (typeof displayName !== 'string' || typeof changeable !== 'boolean') {
        return null;
    }

    return { displayName, changeable };
}
